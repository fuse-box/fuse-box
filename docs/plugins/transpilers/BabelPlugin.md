# BabelPlugin

## Description
The babel plugin is used to transpile code to different dialects of javascript.

## Install
This package depends on the `babel-core` module.

```bash
yarn add babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --dev
npm install babel-core babel-preset-es2015 babel-plugin-transform-react-jsx --save-dev
```

##### Parameters

| Name | Type | Description | ?Default |
| ---- | ---- | ----------- | -------- |
| config | `Object`  | when using other fuse-box only properties, babel config is passed in as .config |  |
| limit2project | `boolean`  | to use this plugin across an entire project (including other modules like npm) | `true` |
| extensions | `Array<string>`  | file extensions to allow with fuse-box | `[".jsx"]`
| test | `Regex`  | files to match | <code>/\\.(j&#124;t)s(x)?$/</code> |




## Examples

### JSX

```js
const {BabelPlugin} = require("fuse-box");

plugins: [
    BabelPlugin({
      config: {
        sourceMaps: true,
        presets: ["es2015"],
        plugins: [
            ["transform-react-jsx"],
        ],
      },
    })
]
```

### Shorthand

when only passing in a babel config, there is no need for .config

```js
plugins: [
  BabelPlugin({
    presets: ["es2015"]
  })
]
```
