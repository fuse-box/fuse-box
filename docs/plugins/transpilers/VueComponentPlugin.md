# VueComponentPlugin

## Description
The VueComponentPlugin is used to transpile `.vue` files into Javascript.

## Install
This plugin depends on the following node modules:
- `vue-template-compiler`
- `vue-template-es2015-compiler`
- `vue-hot-reload-api`
- `postcss-selector-parser`
- `vue`

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler vue-hot-reload-api vue postcss-selector-parser --save-dev
```

For more information about the `*.vue` file specifications please see the official [documentation](https://vue-loader.vuejs.org/en/start/spec.html)

## Usage
Just import from FuseBox:
```js
const { VueComponentPlugin } = require('fuse-box')
```

Inject into a chain:
```js
fuse.plugin(
  VueComponentPlugin()
)
```

Or add it to the main config to make it available across bundles:
```js
FuseBox.init({
  plugins: [
    VueComponentPlugin()
  ]
});
```

## Configuration
### Defaults
If no `lang` attributes are specified (see below) the `VueComponentPlugin` will use the following configuration out of the box:

- `<template>` Defaults to `html` and uses [HTMLPlugin](/plugins/html-plugin)
- `<script>` Defaults to `ts` and uses FuseBox's integrated Typescript compiler
- `<style>` Defaults to `css` and uses [CSSPlugin](/plugins/css-plugin)

### Language Attributes
When `VueComponentPlugin` detects a `lang` attribute on a block it will attempt to match with the corresponding FuseBox plugin. For example:

- `<script lang="coffee">` - [CoffeeScriptPlugin](/plugins/coffeescript-plugin)
- `<script lang="js">` - [BabelPlugin](/plugins/babelplugin) (will read configuration from `.babelrc` file)
- `<style lang="less">` - [LESSPlugin](/plugins/less-plugin)
- `<style lang="scss">` - [SASSPlugin](/plugins/sass-plugin)

note: If using lang="js" and configuration item useTypescriptCompiler: true then FuseBox will use the internal Typescript compiler and NOT BabelPlugin


### Using Custom Plugin Chains
If the above functionality doesn't fit your needs, you can override the pre-processing by optionally setting the `script`, `style` or `template` options. This follows the standard FuseBox way of defining plugin chains:

```js
const { FuseBox, VueComponentPlugin, BabelPlugin, SassPlugin, CSSResourcePlugin, PostCSSPlugin, CSSPlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
      VueComponentPlugin({
          script: BabelPlugin({ ... }),
          template: HTMLPlugin({ ... })
          style: [
              SassPlugin({ ... }),
              CSSResourcePlugin({ ... }),
              PostCSSPlugin( ... ),
              CSSPlugin({ ... })
          ]
      })
    ]
})
```

note: Overriding a plugin chain for a .vue block will make the VueComponentPlugin ignore any lang attributes.

### External Files
The `VueComponentPlugin` fully understands the `src` attribute and will handle external files just the same as inline content.

### Scoped Styling
Scoped styling is fully supported by using the `scoped` attribute. Support for the `module` attribute will be coming soon.

### HMR
Hot Module Reloading is also fully supported, just enable it on your bundles by calling `.hmr()`. See [here](/page/development#hot-module-reload) for more info.
