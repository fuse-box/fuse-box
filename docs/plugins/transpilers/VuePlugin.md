# VuePlugin

## Description
The Vue plugin is used to transpile .vue files into Javascript.

## Install
This package depends on the `vue-template-compiler`, `vue-template-es2015-compiler` and one of `typescript` and `babel-core` modules.

Also, if you want to include Vue library into the bundle, remember to install `vue` module too.

```bash
# Using yarn:
yarn add vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --dev

# Using npm:
npm install vue-template-compiler vue-template-es2015-compiler typescript babel-core vue --save-dev
```

## Example

```js
// fuse.js

const { FuseBox, VuePlugin } = require('fuse-box')

const fsbx = FuseBox.init({
    homeDir: './src',
    output: 'dist/app.js',
    plugins: [
        VuePlugin()
    ]
})

fsbx.dev()

fsbx.bundle('app.js')
    .instructions('>index.ts')
    .watch()

fsbx.run()

```

```typescript
// src/index.ts

import * as Vue from 'vue/dist/vue.js'
import * as App from './components/App.vue'

new Vue({
    el: '#app',
    render: h => h(App)
})

```

```html
<!-- src/components/App.vue -->

<template>
    <div>
        <p>{{ msg }}</p>
        <input type="text" v-model="msg" />
    </div>
</template>

<script>
    export default {
        name: 'app',
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```

## Script transpiler

By default `VuePlugin` uses `typescript` to compile contents of `<script></script>`. By defining the `lang` attribute of the `<script>` tag, content can be compiled using desired transpiler. Currently only `ts` and `babel` are supported.

### Babel

If `babel` attribute is given, VuePlugin will load babel config from there, otherwise VuePlugin will load `.babelrc` from the root of the project. If no config given and no rc found, it uses default config. Packages `babel-core` and `babel-plugin-transform-es2015-modules-commonjs` must be installed to use babel.

```js
// Configuring babel in fuse.js (Default)
VuePlugin({
    babel: {
        config: {
            "plugins": ["transform-es2015-modules-commonjs"]
        }
    }
})

// .babelrc
{
 "plugins": ["transform-es2015-modules-commonjs"]
}
```

## Template engine preprocessor

By defining the `lang` attribute of the `<template>` tag, content can be compiled to HTML using your favorite template engine.

The `consolidate` module must be installed in addition to the template engine module you want to use.

View a list of all [supported template engines](https://github.com/tj/consolidate.js#supported-template-engines).

### Example using Pug:

```bash
# Using yarn:
yarn add consolidate pug --dev

# Using npm:
npm install consolidate pug --save-dev
```

```html
<!-- src/components/App.vue -->

<template lang="pug">
    .content
        p {{ msg }}
        input(type="text", v-model="msg")
</template>

<script>
    export default {
        data () {
            return {
                msg: 'Welcome to Your Vue.js App'
            }
        }
    }
</script>
```
