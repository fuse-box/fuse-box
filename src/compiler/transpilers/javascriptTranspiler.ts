import { ASTNode } from '../interfaces/AST';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { ITranspiler } from '../interfaces/ITranspiler';
import { createGlobalContext } from '../program/GlobalContext';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BundlePolyfillTransformer } from '../transformers/bundle/BundlePolyfillTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ITransformerResult } from '../interfaces/ITranformerResult';

export function javascriptTranspiler(props: ITranspiler): ITransformerResult {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: props.env,
      isBrowser: props.isBrowser,
      isServer: props.isServer,
    }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    props.target === 'browser' && BrowserProcessTransformer({ env: props.env, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...props, onRequireCallExpression }),

    DynamicImportTransformer({ onRequireCallExpression }),

    JSXTransformer(),

    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  if (props.transformers && props.transformers.length) {
    const opts = { module: props.module, onRequireCallExpression };
    for (const transformer of props.transformers) {
      defaultTransformers.unshift(transformer(opts));
    }
  }

  transpileModule({
    ast: props.ast as ASTNode,
    globalContext: createGlobalContext({}),
    transformers: defaultTransformers,
  });

  return { ast: props.ast, requireStatementCollection };
}
