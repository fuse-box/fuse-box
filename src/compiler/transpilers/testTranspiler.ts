import * as fs from 'fs';
import * as path from 'path';
import { ITarget } from '../../config/PrivateConfig';
import { createContext } from '../../core/Context';
import { createModule } from '../../core/Module';
import { createPackage } from '../../core/Package';
import { makeFuseBoxPath } from '../../utils/utils';
import { generate } from '../generator/generator';
import { ASTNode } from '../interfaces/AST';
import { parseTypeScript, parseJavascript } from '../parser';
import { transformCommonVisitors } from '../transformer';

export interface ICompileModuleProps {
  target?: ITarget;
  env?: { [key: string]: string };
  fileName?: string;
  code?: string;
  emitDecoratorMetadata?: boolean;
  useMeriyah?: boolean;
}

export function testTranspile(props: ICompileModuleProps) {
  let contents = props.code;

  if (!contents) {
    contents = fs.readFileSync(props.fileName).toString();
  }

  const ext = props.fileName ? path.extname(props.fileName) : '.tsx';

  let ast;
  if (props.useMeriyah) {
    ast = parseJavascript(contents, { jsx: true });
  } else {
    ast = parseTypeScript(contents, { jsx: true });
  }

  const ctx = createContext({
    cache: false,
    devServer: false,
    target: props.target || 'browser',
    env: props.env,
  });

  if (props.emitDecoratorMetadata) {
    ctx.tsConfig.compilerOptions.emitDecoratorMetadata = props.emitDecoratorMetadata;
  }

  const pkg = createPackage({ ctx: ctx, meta: {} as any });
  const module = createModule(
    {
      absPath: __filename,
      ctx: ctx,
      extension: ext,
      fuseBoxPath: makeFuseBoxPath(path.dirname(props.fileName || __filename), props.fileName || __filename),
    },
    pkg,
  );
  module.ast = ast as ASTNode;

  transformCommonVisitors(module);

  const res = generate(ast, {});

  return res;
}
