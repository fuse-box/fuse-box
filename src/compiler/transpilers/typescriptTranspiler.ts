import { ASTNode } from '../interfaces/AST';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { ITranspiler } from '../interfaces/ITranspiler';
import { createGlobalContext } from '../program/GlobalContext';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';
import { BundlePolyfillTransformer } from '../transformers/bundle/BundlePolyfillTransformer';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ClassConstructorPropertyTransformer } from '../transformers/ts/ClassConstructorPropertyTransformer';
import { CommonTSfeaturesTransformer } from '../transformers/ts/CommonTSfeaturesTransformer';
import { EnumTransformer } from '../transformers/ts/EnumTransformer';
import { NamespaceTransformer } from '../transformers/ts/NameSpaceTransformer';
import { ITransformerResult } from '../interfaces/ITranformerResult';
import { DecoratorTransformer } from '../transformers/ts/decorators/DecoratorTransformer';

export function typescriptTranspiler(props: ITranspiler): ITransformerResult {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: props.env,
      isBrowser: props.isBrowser,
      isServer: props.isServer,
    }),
    DecoratorTransformer({ onRequireCallExpression, emitDecoratorMetadata: props.emitDecoratorMetadata }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    props.target === 'browser' && BrowserProcessTransformer({ env: props.env, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...props, onRequireCallExpression }),

    DynamicImportTransformer({ onRequireCallExpression }),
    EnumTransformer(),
    ClassConstructorPropertyTransformer(),
    JSXTransformer(),
    NamespaceTransformer(),

    CommonTSfeaturesTransformer(),
    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  if (props.transformers && props.transformers.length) {
    const opts = { module: props.module, onRequireCallExpression };
    for (const transformer of props.transformers) {
      defaultTransformers.unshift(transformer(opts));
    }
  }

  transpileModule({
    ast: props.ast as ASTNode,
    globalContext: createGlobalContext({}),
    transformers: defaultTransformers,
  });

  return { ast: props.ast, requireStatementCollection };
}
