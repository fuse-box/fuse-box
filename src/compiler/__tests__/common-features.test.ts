import { testTranspile } from '../transpilers/testTranspiler';

describe('Common features test', () => {
  describe('Abstract methods', () => {
    it('should remove abstract methods', () => {
      const result = testTranspile({
        code: `
          abstract  class A {
            constructor(public name){}
            abstract hello()
        }

      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Should remove declare', () => {
    it('should remove consts', () => {
      const result = testTranspile({
        code: `
        alert(1)
        declare const A: any, B: any;
        alert(2)
      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Interface removal', () => {
    it('should remove export interface', () => {
      const result = testTranspile({
        code: `
          console.log(1);
          export interface Foo {

          }
          console.log(2);
      `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should remove interface', () => {
      const result = testTranspile({
        code: `
           alert(1);
           interface Foo {}
           alert(2);
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should ignore export declare', () => {
      const result = testTranspile({
        code: `
        alert(1)
        export declare const A: any, B: any, oi = 2;
        alert(2)
        `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });
});
