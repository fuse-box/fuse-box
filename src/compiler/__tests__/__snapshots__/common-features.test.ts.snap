// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common features test Abstract methods should remove abstract methods 1`] = `
"class A {
  constructor(name) {
    this.name = name;
  }
}
"
`;

exports[`Common features test Interface removal should ignore export declare 1`] = `
"alert(1);
alert(2);
"
`;

exports[`Common features test Interface removal should ignore export declare 2`] = `
"alert(1);
"
`;

exports[`Common features test Interface removal should remove export interface 1`] = `
"console.log(1);
console.log(2);
"
`;

exports[`Common features test Interface removal should remove interface 1`] = `
"alert(1);
alert(2);
"
`;

exports[`Common features test Should remove declare should remove consts 1`] = `
"alert(1);
alert(2);
"
`;

exports[`Common features test Various should handle NotNullExpression 1`] = `
"function hey(res) {
  return res;
}
"
`;

exports[`Common features test Various should remove type assertion 1`] = `
"if (_e.target === 1) {}
"
`;
